thermal.c

```diff
@@ -14,43 +14,333 @@
+		policy);	/* validation policy */
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: group id - failed to create attributes.\n");
+		return NL_SKIP;
+	}
+
+	/* start of the multi-cast group attribute */
+	mcgrp = attrs[CTRL_ATTR_MCAST_GROUPS];
+	if (!mcgrp) {
+		log(TO_ALL, LOG_ERR, "thermal: group id - no multi-cast group attributes.\n");
+		return NL_SKIP;
+	}
+
+	/* iterate a stream of nested attributes to get the group id */
+	nla_for_each_nested(cur_mcgrp, mcgrp, i) {
+		struct nlattr *nested_attrs[CTRL_ATTR_MCAST_GRP_MAX + 1];
+		struct nlattr *name, *id;
+
+		/* get start and length of payload section */
+		attrhdr = nla_data(cur_mcgrp);
+		attrlen = nla_len(cur_mcgrp);
+
+		rc = nla_parse(nested_attrs, CTRL_ATTR_MCAST_GRP_MAX, attrhdr, attrlen, policy);
+		if (rc)
+			continue;
+
+		name = nested_attrs[CTRL_ATTR_MCAST_GRP_NAME];
+		id = nested_attrs[CTRL_ATTR_MCAST_GRP_ID];
+		if (!name || !id)
+			continue;
+
+		if (strncmp(nla_data(name), data->group, nla_len(name)) != 0)
+			continue;
+
+		data->id = nla_get_u32(id);
+		log(TO_ALL, LOG_DEBUG, "thermal: received group id (%d).\n", data->id);
+		break;
+	}
+	return NL_OK;
+}
+
+static int handle_error(struct sockaddr_nl *sk_addr __attribute__((unused)),
+			struct nlmsgerr *err, void *arg)
+{
+	if (arg) {
+		log(TO_ALL, LOG_INFO, "thermal: received a netlink error (%s).\n",
+		    nl_geterror(err->error));
+		*((int *)arg) = err->error;
+	}
+	return NL_SKIP;
+}
+
+static int handle_end(struct nl_msg *msg __attribute__((unused)), void *arg)
+{
+	if (arg)
+		*((int *)arg) = 0;
+	return NL_SKIP;
+}
+
+#define NL_FAMILY_NAME  "nlctrl"
+
+struct msgheader {
+	unsigned char cmd, version;
+	unsigned int port, seq;
+	int id, hdrlen, flags;
+};
 
 static gboolean establish_netlink(void)
 {
-	gboolean error = TRUE;
+	struct msgheader msghdr = { CTRL_CMD_GETFAMILY, 0, 0, 0, 0, 0, 0 };
+	struct family_data nldata = { THERMAL_GENL_EVENT_GROUP_NAME, -ENOENT };
+	struct nl_cb *cloned_callback = NULL;
+	int rc, group_id, callback_rc = 1;
+	struct nl_msg *msg = NULL;
+	gboolean error = FALSE;
+	void *hdr;
+
+	msg = nlmsg_alloc();
+	if (!msg) {
+		log(TO_ALL, LOG_ERR, "thermal: message allocation failed.\n");
+		goto out;
+	}
+
+	msghdr.id = genl_ctrl_resolve(sock, NL_FAMILY_NAME);
+	if (msghdr.id < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: message id enumeration failed.\n");
+		goto err_out;
+	}
+
+	hdr = genlmsg_put(msg, msghdr.port, msghdr.seq, msghdr.id, msghdr.hdrlen,
+			  msghdr.flags, msghdr.cmd, msghdr.version);
+	if (!hdr) {
+		log(TO_ALL, LOG_ERR, "thermal: netlink header setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nla_put_string(msg, CTRL_ATTR_FAMILY_NAME, THERMAL_GENL_FAMILY_NAME);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: message setup failed.\n");
+		goto err_out;
+	}
+
+	cloned_callback = nl_cb_clone(callback);
+	if (!cloned_callback) {
+		log(TO_ALL, LOG_ERR, "thermal: callback handle duplication failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_send_auto_complete(sock, msg);
```

Comment:
FYI, not a huge deal but nl_send_auto_complete is deprecated in libnl3, probably want to use nl_send_auto instead

Reply:
this is still deprecated

Reply:
Will do. Sorry for missing this too.

Reply:
Updated: https://github.com/Irqbalance/irqbalance/compare/3ec98d963fa3a483073e73a73998c2f317dabfc5..560291389e33db108d2fb6d954ef059f953c6e33#diff-3066472ef74ce9f800762e2831b21aea1cd51c554c719742f0ece2f74c2e4a03L256